### PR TITLE
python: tweaks in preparation for flux-tree-helper

### LINKED ISSUE #2535: ci: pylint does not run against python files in `src/cmd`
When moving code from `src/cmd/flux-mini` to `src/bindings/python/job.py`, I realized that pylint only runs against the code in `src/bindings/python`.  We should probably have it run in both.

### PR DESCRIPTION
**Pull request by SteVwonder**
- Add `job_kvs` and `job_kvs_guest` bindings.
- Add a `attr_get` method to the `Flux` class.
- Make `OutputFormat` publicly accessible.
  - Add some more tests of `--format` to ensure the move didn't break anything
- Two misc fixes in `t/python`.
- run pylint on python scripts in `src/cmd`
- appease pylint
- automatically set `FLUX_PYCLI_LOGLEVEL=10` when the `debug` variable is set during `make check`

Closes #2535

### CODE REVIEW

**Comment by grondo** [comment]
> I don't think I broke anything with flux jobs in the process of generalizing the OutputFormat class. Any thoughts on how to check for possible regressions?

There are quite a few tests of `flux jobs`, as long as none of those fail it is a good start.

Off the top of my head, a few things I'm not sure are in the current tests:
 - a sensible error is printed on an invalid format
 - formats for flux-jobs works with `{0.key}` and `{key}`
 - errors thrown when a format spec is invalid `{key:garbage}`

**Comment by grondo** [comment]
I guess the python linter didn't run on `src/cmd/flux-jobs.py` before, and it doesn't like the single letter variables, sorry! In this case I can't agree with the lint result, sometimes a single letter variable is just more readable, but oh well!

**Comment by SteVwonder** [comment]
>   a sensible error is printed on an invalid format
    formats for flux-jobs works with {0.key} and {key}
    errors thrown when a format spec is invalid {key:garbage}

Excellent suggestions, thanks!  I'll add some CI tests for errors on invalid formats.  Should I add a test for the `{0.key}` case, or is testing the output of a command is too fragile of a test?

> In this case I can't agree with the lint result, sometimes a single letter variable is just more readable, but oh well!

Yeah, I agree.  I'll put linter exceptions in the comments where appropriate and lengthen the variable names otherwise.

**Comment by grondo** [comment]
> Should I add a test for the {0.key} case, or is testing the output of a command is too fragile of a test?

I think as long as the command doesn't unexpectedly fail, that is a fine test.

For flux-jobs specifically though, we have the `--from-stdin` option that allows you to capture and replay `job_list` results for reproducible output...

**Comment by SteVwonder** [comment]
Tests pass for me under python3, but are failing under python2.  I did a pass to take care of the low-hanging fruit, but there are a few things left that are just painful enough that I'm tempted to punt on them and just wait for #2805 to get merged.

**Comment by chu11** [comment]
Minor comment, in PR #2798, i use the OutputFormat's `.fields` object member.

When OutputFormat was private in `flux-jobs`, no big deal.  But now that it's public in `util.py`, is it considered kosher for code to use `obj.fields`?  Should there be a getter function?  Or perhaps some documentation that says "these class variables are public"?

**Comment by SteVwonder** [comment]
Thanks for the feedback @chu11.

> But now that it's public in util.py, is it considered kosher for code to use obj.fields?

Generally, I think that it is kosher for code to use any object attributes that aren't prefixed by an underscore. [[1]](https://docs.python.org/2/tutorial/classes.html#tut-private)

> Or perhaps some documentation that says "these class variables are public"?

Today I Learned (TIL) Sphinx supports that type of documentation via the `ivar` info field [[2]](https://stackoverflow.com/questions/8649105/how-to-show-instance-attributes-in-sphinx-doc)

> Should there be a getter function?

Do we ever expect users to change the `fields` value?  If not, I think a getter function in the form of a `property` is the right way to go to avoid accidental overwriting.  I pushed a presumptive `fixup` commit with that change.

**Comment by chu11** [comment]
> Do we ever expect users to change the fields value? If not, I think a getter function in the form of a property is the right way to go to avoid accidental overwriting. I pushed a presumptive fixup commit with that change.

LGTM!

**Comment by SteVwonder** [comment]
@mergifyio rebase

**Comment by grondo** [comment]
chain lint error in the `t2800-jobs-cmd.t` test
```
error: bug in the test script: broken &&-chain: 
        flux jobs --state=pending --format="{runtime:0.3f}" > runtime-header.out &&
        echo "RUNTIME" >> runtime-header.exp &&
        for i in `seq 1 6`; do echo "0.000" >> runtime-header.exp; done
        test_cmp runtime-header.out runtime-header.exp
```

**Comment by grondo** [comment]
I think I saw another hang here. Tests that didn't report seemed to be
```diff
--- good.output	2020-03-14 18:48:55.491293717 +0000
+++ hang.output	2020-03-14 18:49:04.251351552 +0000
@@ -9,12 +9,10 @@
 lua/t1004-statwatcher.t
 python/t0001-handle.py
 python/t0002-wrapper.py
-python/t0003-barrier.py
 python/t0004-event.py
 python/t0005-kvs.py
 python/t0006-request.py
 python/t0007-watchers.py
-python/t0009-security.py
 python/t0010-job.py
 python/t0012-futures.py
 python/t0013-job-list-inactive.py

```

**Comment by SteVwonder** [comment]
Rebased on top of the latest upstream master (yay for py3 only!), got pylint running on the scripts in `src/cmd` and then addressed the pylint fallout from the `src/cmd` addition and the move to py3. I think this PR is in a good spot now and is ready for another review.

**Comment by SteVwonder** [comment]
> A little difficult to review this because of all the pedantic changes required by pylint

Yeah, sorry about that.  I tried to contain them to their own commits, so the other changes could be isolated.  Any suggestions on a better way to approach it in the future?

File: configure.ac [hunk 1]
```diff
@@ -316,9 +316,9 @@ AS_IF([test "x$enable_pylint" = "xyes"], [
   AC_CHECK_PROG(PYLINT,[pylint],[pylint])
   AS_IF([test "x$PYLINT" != "xpylint"], [AC_MSG_ERROR([No pylint found in PATH])])
   AM_CHECK_PYMOD(pylint,
-                   [StrictVersion(pylint.__version__) >= StrictVersion('1.4.5')],
+                   [StrictVersion(pylint.__version__) >= StrictVersion('1.8.4')],
                    ,
-                   [AC_MSG_ERROR([could not find python module pylint, version 1.4.5+ required])]
+                   [AC_MSG_ERROR([could not find python module pylint, version 1.8.4+ required])]
```
[the remark this answers is not in the document]
Comment (SteVwonder) [reply configure.ac]: Yeah. That's a bummer.  I originally tried earlier versions (even 1.6.5); unfortunately 1.7.0 is the earliest pylint that supports python 3 semantics, and 1.8.X is the earliest that doesn't have a false positive `cyclic-import` error on our code-base.  Which reminds me, we should probably have a `requirements.txt` file to make installing our python requirements via pip easier.  I'll open an issue on that.

**Comment by grondo** [comment]
> Any suggestions on a better way to approach it in the future?

No I think you did fine. I just needed an outlet to complain about pylint's punctilious pettifoggery!

I think we can merge this one since you've got a LGTM comment from @chu11 as well. I'll set MWP.

**Pull request merged**